File: runners/js/features/step_definitions/request.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { TestWorld } from './world';

// Header validation step definitions for JavaScript SDK

When('I set the request header {string} to {string}', function(headerName, headerValue) {
  if (!this.requestHeaders) {
    this.requestHeaders = {};
  }
  this.requestHeaders[headerName] = headerValue;
});

When('I clear the request header {string}', function(headerName) {
  if (this.requestHeaders) {
    delete this.requestHeaders[headerName];
  }
});

Then('the request should have included header {string} with value {string}', function(headerName, expectedValue) {
  if (!this.capturedRequestHeaders) {
    throw new Error('No request headers were captured');
  }

  // Check for header (case-insensitive)
  const actualValue = this.findHeaderValue(this.capturedRequestHeaders, headerName);
  if (actualValue === undefined) {
    throw new Error(`Request header '${headerName}' was not found`);
  }

  if (actualValue !== expectedValue) {
    throw new Error(`Request header '${headerName}' expected value '${expectedValue}', got '${actualValue}'`);
  }
});

Then('the request should have included header {string}', function(headerName) {
  if (!this.capturedRequestHeaders) {
    throw new Error('No request headers were captured');
  }

  const actualValue = this.findHeaderValue(this.capturedRequestHeaders, headerName);
  if (actualValue === undefined) {
    throw new Error(`Request header '${headerName}' was not found`);
  }
});

Then('the request should not have included header {string}', function(headerName) {
  if (!this.capturedRequestHeaders) {
    return; // No headers captured means header wasn't included
  }

  const actualValue = this.findHeaderValue(this.capturedRequestHeaders, headerName);
  if (actualValue !== undefined) {
    throw new Error(`Request header '${headerName}' was found but should not have been included`);
  }
});

Then('the response should include header {string} with value {string}', function(headerName, expectedValue) {
  if (!this.lastResponse) {
    throw new Error('No response received');
  }

  const responseHeaders = this.extractResponseHeaders();
  if (!responseHeaders) {
    throw new Error('No response headers available');
  }

  const actualValue = this.findHeaderValue(responseHeaders, headerName);
  if (actualValue === undefined) {
    throw new Error(`Response header '${headerName}' was not found`);
  }

  if (actualValue !== expectedValue) {
    throw new Error(`Response header '${headerName}' expected value '${expectedValue}', got '${actualValue}'`);
  }
});

Then('the authorization header should contain {string}', function(expectedSubstring) {
  if (!this.capturedRequestHeaders) {
    throw new Error('No request headers were captured');
  }

  const authValue = this.findHeaderValue(this.capturedRequestHeaders, 'Authorization');
  if (authValue === undefined) {
    throw new Error('Authorization header was not found');
  }

  if (!authValue.includes(expectedSubstring)) {
    throw new Error(`Authorization header '${authValue}' does not contain '${expectedSubstring}'`);
  }
});

Then('the content length should be greater than {int}', function(minLength) {
  if (!this.capturedRequestHeaders) {
    throw new Error('No request headers were captured');
  }

  const contentLengthStr = this.findHeaderValue(this.capturedRequestHeaders, 'Content-Length');
  if (contentLengthStr === undefined) {
    throw new Error('Content-Length header was not found');
  }

  const contentLength = parseInt(contentLengthStr, 10);
  if (isNaN(contentLength)) {
    throw new Error(`Invalid Content-Length value: ${contentLengthStr}`);
  }

  if (contentLength <= minLength) {
    throw new Error(`Content-Length ${contentLength} is not greater than ${minLength}`);
  }
});

Then('both responses should be successful', function() {
  if (!this.responseHistory || this.responseHistory.length < 2) {
    throw new Error(`Expected at least 2 responses, got ${this.responseHistory ? this.responseHistory.length : 0}`);
  }

  // Check last two responses
  const lastTwo = this.responseHistory.slice(-2);
  for (let i = 0; i < lastTwo.length; i++) {
    if (lastTwo[i].error) {
      throw new Error(`Response ${i + 1} failed: ${lastTwo[i].error.message}`);
    }
  }
});

Then('both requests should have included header {string} with value {string}', function(headerName, expectedValue) {
  if (!this.requestHeaderHistory || this.requestHeaderHistory.length < 2) {
    throw new Error(`Expected at least 2 request header sets, got ${this.requestHeaderHistory ? this.requestHeaderHistory.length : 0}`);
  }

  // Check last two request header sets
  const lastTwo = this.requestHeaderHistory.slice(-2);
  for (let i = 0; i < lastTwo.length; i++) {
    const actualValue = this.findHeaderValue(lastTwo[i], headerName);
    if (actualValue === undefined) {
      throw new Error(`Request ${i + 1} header '${headerName}' was not found`);
    }
    if (actualValue !== expectedValue) {
      throw new Error(`Request ${i + 1} header '${headerName}' expected value '${expectedValue}', got '${actualValue}'`);
    }
  }
});

Then('the first request should have included header {string} with value {string}', function(headerName, expectedValue) {
  if (!this.requestHeaderHistory || this.requestHeaderHistory.length < 1) {
    throw new Error('No request headers in history');
  }

  const firstHeaders = this.requestHeaderHistory[0];
  const actualValue = this.findHeaderValue(firstHeaders, headerName);
  if (actualValue === undefined) {
    throw new Error(`First request header '${headerName}' was not found`);
  }
  if (actualValue !== expectedValue) {
    throw new Error(`First request header '${headerName}' expected value '${expectedValue}', got '${actualValue}'`);
  }
});

Then('the second request should not have included header {string}', function(headerName) {
  if (!this.requestHeaderHistory || this.requestHeaderHistory.length < 2) {
    throw new Error(`Expected at least 2 request header sets, got ${this.requestHeaderHistory ? this.requestHeaderHistory.length : 0}`);
  }

  const secondHeaders = this.requestHeaderHistory[1];
  const actualValue = this.findHeaderValue(secondHeaders, headerName);
  if (actualValue !== undefined) {
    throw new Error(`Second request header '${headerName}' was found but should not have been included`);
  }
});

When('I make a preflight request to Check endpoint', function() {
  // Placeholder for CORS preflight request
  throw new Error('Preflight requests not yet implemented in JavaScript SDK');
});

Then('the preflight response should be successful', function() {
  // Placeholder for preflight response validation
  throw new Error('Preflight response validation not yet implemented');
});

// Helper methods (these would be added to the test context)

// Add to World/TestContext prototype
if (typeof module !== 'undefined' && module.exports) {
  // Helper functions to be added to test context
  const headerHelpers = {
    findHeaderValue: function(headers, headerName) {
      if (!headers) return undefined;
      
      // Check for exact match first
      if (headers[headerName] !== undefined) {
        return headers[headerName];
      }
      
      // Case-insensitive search
      const lowerHeaderName = headerName.toLowerCase();
      for (const [name, value] of Object.entries(headers)) {
        if (name.toLowerCase() === lowerHeaderName) {
          return value;
        }
      }
      
      return undefined;
    },

    extractResponseHeaders: function() {
      if (!this.lastResponse) return null;
      
      // Extract headers from response object
      // Implementation depends on the JavaScript SDK response structure
      if (this.lastResponse.headers) {
        return this.lastResponse.headers;
      }
      
      // If response has a headers property in different format
      if (this.lastResponse._headers) {
        return this.lastResponse._headers;
      }
      
      // Return stored response headers if available
      return this.lastResponseHeaders || {};
    },

    captureRequestHeaders: function(requestConfig) {
      if (!this.capturedRequestHeaders) {
        this.capturedRequestHeaders = {};
      }
      
      // Capture headers from request configuration
      if (requestConfig && requestConfig.headers) {
        Object.assign(this.capturedRequestHeaders, requestConfig.headers);
      }
      
      // Add custom request headers
      if (this.requestHeaders) {
        Object.assign(this.capturedRequestHeaders, this.requestHeaders);
      }
      
      // Store in history
      if (!this.requestHeaderHistory) {
        this.requestHeaderHistory = [];
      }
      this.requestHeaderHistory.push({ ...this.capturedRequestHeaders });
    },

    storeResponseInHistory: function(response, error) {
      if (!this.responseHistory) {
        this.responseHistory = [];
      }
      this.responseHistory.push({ response, error });
    }
  };

  module.exports = headerHelpers;
}

File: runners/python/steps/request.py
from behave import given, when, then
from assertpy import assert_that

# Header validation step definitions for Python SDK

@when('I set the request header "{header_name}" to "{header_value}"')
def step_set_request_header(context, header_name, header_value):
    if not hasattr(context.test_context, 'request_headers'):
        context.test_context.request_headers = {}
    context.test_context.request_headers[header_name] = header_value

@when('I clear the request header "{header_name}"')
def step_clear_request_header(context, header_name):
    if hasattr(context.test_context, 'request_headers') and context.test_context.request_headers:
        context.test_context.request_headers.pop(header_name, None)

@then('the request should have included header "{header_name}" with value "{expected_value}"')
def step_request_should_have_header_with_value(context, header_name, expected_value):
    if not hasattr(context.test_context, 'captured_request_headers') or not context.test_context.captured_request_headers:
        raise AssertionError('No request headers were captured')
    
    actual_value = find_header_value(context.test_context.captured_request_headers, header_name)
    if actual_value is None:
        raise AssertionError(f"Request header '{header_name}' was not found")
    
    if actual_value != expected_value:
        raise AssertionError(f"Request header '{header_name}' expected value '{expected_value}', got '{actual_value}'")

@then('the request should have included header "{header_name}"')
def step_request_should_have_header(context, header_name):
    if not hasattr(context.test_context, 'captured_request_headers') or not context.test_context.captured_request_headers:
        raise AssertionError('No request headers were captured')
    
    actual_value = find_header_value(context.test_context.captured_request_headers, header_name)
    if actual_value is None:
        raise AssertionError(f"Request header '{header_name}' was not found")

@then('the request should not have included header "{header_name}"')
def step_request_should_not_have_header(context, header_name):
    if not hasattr(context.test_context, 'captured_request_headers') or not context.test_context.captured_request_headers:
        return  # No headers captured means header wasn't included
    
    actual_value = find_header_value(context.test_context.captured_request_headers, header_name)
    if actual_value is not None:
        raise AssertionError(f"Request header '{header_name}' was found but should not have been included")

@then('the response should include header "{header_name}" with value "{expected_value}"')
def step_response_should_include_header_with_value(context, header_name, expected_value):
    if not context.test_context.last_response:
        raise AssertionError('No response received')
    
    response_headers = extract_response_headers(context.test_context)
    if not response_headers:
        raise AssertionError('No response headers available')
    
    actual_value = find_header_value(response_headers, header_name)
    if actual_value is None:
        raise AssertionError(f"Response header '{header_name}' was not found")
    
    if actual_value != expected_value:
        raise AssertionError(f"Response header '{header_name}' expected value '{expected_value}', got '{actual_value}'")

@then('the response should include header "{header_name}"')
def step_response_should_include_header(context, header_name):
    if not context.test_context.last_response:
        raise AssertionError('No response received')
    
    response_headers = extract_response_headers(context.test_context)
    if not response_headers:
        raise AssertionError('No response headers available')
    
    actual_value = find_header_value(response_headers, header_name)
    if actual_value is None:
        raise AssertionError(f"Response header '{header_name}' was not found")

@then('the authorization header should contain "{expected_substring}"')
def step_authorization_header_should_contain(context, expected_substring):
    if not hasattr(context.test_context, 'captured_request_headers') or not context.test_context.captured_request_headers:
        raise AssertionError('No request headers were captured')
    
    auth_value = find_header_value(context.test_context.captured_request_headers, 'Authorization')
    if auth_value is None:
        raise AssertionError('Authorization header was not found')
    
    if expected_substring not in auth_value:
        raise AssertionError(f"Authorization header '{auth_value}' does not contain '{expected_substring}'")

@then('the content length should be greater than {min_length:d}')
def step_content_length_should_be_greater_than(context, min_length):
    if not hasattr(context.test_context, 'captured_request_headers') or not context.test_context.captured_request_headers:
        raise AssertionError('No request headers were captured')
    
    content_length_str = find_header_value(context.test_context.captured_request_headers, 'Content-Length')
    if content_length_str is None:
        raise AssertionError('Content-Length header was not found')
    
    try:
        content_length = int(content_length_str)
    except ValueError:
        raise AssertionError(f"Invalid Content-Length value: {content_length_str}")
    
    if content_length <= min_length:
        raise AssertionError(f"Content-Length {content_length} is not greater than {min_length}")

@then('both responses should be successful')
def step_both_responses_should_be_successful(context):
    if not hasattr(context.test_context, 'response_history') or len(context.test_context.response_history) < 2:
        response_count = len(context.test_context.response_history) if hasattr(context.test_context, 'response_history') else 0
        raise AssertionError(f"Expected at least 2 responses, got {response_count}")
    
    # Check last two responses
    last_two = context.test_context.response_history[-2:]
    for i, response_record in enumerate(last_two):
        if response_record.get('error'):
            raise AssertionError(f"Response {i + 1} failed: {response_record['error']}")

@then('both requests should have included header "{header_name}" with value "{expected_value}"')
def step_both_requests_should_have_header_with_value(context, header_name, expected_value):
    if not hasattr(context.test_context, 'request_header_history') or len(context.test_context.request_header_history) < 2:
        history_count = len(context.test_context.request_header_history) if hasattr(context.test_context, 'request_header_history') else 0
        raise AssertionError(f"Expected at least 2 request header sets, got {history_count}")
    
    # Check last two request header sets
    last_two = context.test_context.request_header_history[-2:]
    for i, headers in enumerate(last_two):
        actual_value = find_header_value(headers, header_name)
        if actual_value is None:
            raise AssertionError(f"Request {i + 1} header '{header_name}' was not found")
        if actual_value != expected_value:
            raise AssertionError(f"Request {i + 1} header '{header_name}' expected value '{expected_value}', got '{actual_value}'")

@then('the first request should have included header "{header_name}" with value "{expected_value}"')
def step_first_request_should_have_header_with_value(context, header_name, expected_value):
    if not hasattr(context.test_context, 'request_header_history') or len(context.test_context.request_header_history) < 1:
        raise AssertionError('No request headers in history')
    
    first_headers = context.test_context.request_header_history[0]
    actual_value = find_header_value(first_headers, header_name)
    if actual_value is None:
        raise AssertionError(f"First request header '{header_name}' was not found")
    if actual_value != expected_value:
        raise AssertionError(f"First request header '{header_name}' expected value '{expected_value}', got '{actual_value}'")

@then('the second request should not have included header "{header_name}"')
def step_second_request_should_not_have_header(context, header_name):
    if not hasattr(context.test_context, 'request_header_history') or len(context.test_context.request_header_history) < 2:
        history_count = len(context.test_context.request_header_history) if hasattr(context.test_context, 'request_header_history') else 0
        raise AssertionError(f"Expected at least 2 request header sets, got {history_count}")
    
    second_headers = context.test_context.request_header_history[1]
    actual_value = find_header_value(second_headers, header_name)
    if actual_value is not None:
        raise AssertionError(f"Second request header '{header_name}' was found but should not have been included")

@when('I make a preflight request to Check endpoint')
def step_make_preflight_request(context):
    # Placeholder for CORS preflight request
    raise NotImplementedError('Preflight requests not yet implemented in Python SDK')

@then('the preflight response should be successful')
def step_preflight_response_should_be_successful(context):
    # Placeholder for preflight response validation
    raise NotImplementedError('Preflight response validation not yet implemented')

# Helper functions

def find_header_value(headers, header_name):
    """Find header value with case-insensitive matching"""
    if not headers:
        return None
    
    # Check for exact match first
    if header_name in headers:
        return headers[header_name]
    
    # Case-insensitive search
    lower_header_name = header_name.lower()
    for name, value in headers.items():
        if name.lower() == lower_header_name:
            return value
    
    return None

def extract_response_headers(test_context):
    """Extract headers from response object"""
    if not test_context.last_response:
        return None
    
    # Check if response has headers attribute
    if hasattr(test_context.last_response, 'headers'):
        return test_context.last_response.headers
    
    # Check if response has _headers attribute
    if hasattr(test_context.last_response, '_headers'):
        return test_context.last_response._headers
    
    # Return stored response headers if available
    if hasattr(test_context, 'last_response_headers'):
        return test_context.last_response_headers
    
    return {}

def capture_request_headers(test_context, request_config=None):
    """Capture request headers for validation"""
    if not hasattr(test_context, 'captured_request_headers'):
        test_context.captured_request_headers = {}
    
    # Capture headers from request configuration
    if request_config and hasattr(request_config, 'headers'):
        test_context.captured_request_headers.update(request_config.headers)
    
    # Add custom request headers
    if hasattr(test_context, 'request_headers') and test_context.request_headers:
        test_context.captured_request_headers.update(test_context.request_headers)
    
    # Store in history
    if not hasattr(test_context, 'request_header_history'):
        test_context.request_header_history = []
    test_context.request_header_history.append(dict(test_context.captured_request_headers))

def store_response_in_history(test_context, response, error=None):
    """Store response in history for multi-response validation"""
    if not hasattr(test_context, 'response_history'):
        test_context.response_history = []
    test_context.response_history.append({'response': response, 'error': error})
